3DAmsterdam/Assets/Amsterdam3D/Scripts/Interface/Snapshot.cs

```diff
@@ -15,24 +20,241 @@ public class Snapshot : MonoBehaviour
+                
+                // Creates off-screen render texture that can be rendered into
+                screenshotRenderTexture = new RenderTexture(width, height, 24);
+                screenShot = new Texture2D(width, height, TextureFormat.RGB24, false);
+
+                snapshotCamera.targetTexture = screenshotRenderTexture;
+                RenderTexture.active = screenshotRenderTexture;
+
+                // Calls events on the camera related to rendering
+                snapshotCamera.Render();          
+                
+                // Resets canvas
+                responsiveCanvas.worldCamera = null;
+                responsiveCanvas.renderMode = RenderMode.ScreenSpaceOverlay;
+
+                screenShot.ReadPixels(new Rect(0, 0, width, height), 0, 0);
+                screenShot.Apply();
+
+                // Resets variables
+                snapshotCamera.targetTexture = null;
+                RenderTexture.active = null;
+                snapshotMainMenu.isOn = true;
+
+
+                // If no filetype is given, make it a png
+                if (fileType == "")
+                {
+                    fileType = ".png";
+                }
+
+                // Default filename
+                if (fileName == "")
+                {
+                    fileName = "Screenshot_" + DateTime.Now.ToString("yyyy-MM-dd_HH-mm-ss") + "." + fileType;
+                }
+                else
+                {
+                    fileName = fileName + "." + fileType;
+                }
+
+                byte[] bytes = null;
+                if (fileType == "png")
+                {
+                    bytes = screenShot.EncodeToPNG();
+                }
+                else if (fileType == "jpg")
+                {
+                    bytes = screenShot.EncodeToJPG();
+                }
+                else if (fileType == "raw")
+                {
+                    bytes = screenShot.GetRawTextureData();
+                }
+
+                DownloadFile(bytes, bytes.Length, fileName);
+
+                // Exits out of loop
+                StopCoroutine(screenshotCoroutine);
+                gameObject.SetActive(false);
+            }
+        }
+
+    }
+
+
+#if UNITY_EDITOR
+
+    private IEnumerator ScreenshottingEditor()
+    {
+        while(true)
+        {
+            // Helps with making sure the camera is ready to render
+            yield return new WaitForEndOfFrame();
+            // A trick to make sure the UI gets rendered to the screenshot
+            yield return new WaitForEndOfFrame();
+            if (takeScreenshotOnNextFrame)
+            {
+                takeScreenshotOnNextFrame = false;
+
+                // Creates off-screen render texture that can be rendered into
+                screenshotRenderTexture = new RenderTexture(width, height, 24);
+                screenShot = new Texture2D(width, height , TextureFormat.RGB24, false);
```

Comment:
Try to avoid duplication as much as you can. you might have already noticed that you had to make same changes twice, because both methods (for editor and webgl) share a lot of logic.

In VS you can easily move lines into a method by for example selecting line 106 to 126, then do CTRL+R, CTRL+M
and name it to something clear what that part of the code does. For example RenderCameraToTexture()
You can then use that method in both the webgl as the unity editor methods

Reply:
I totally agree, it was something that I was starting to do when working on the new UI.